internal/venice-common/src/main/java/com/linkedin/venice/pubsub/api/PubSubProducerCallback.java

```diff
@@ -1,12 +1,15 @@
 package com.linkedin.venice.pubsub.api;
 
+import java.util.concurrent.CompletableFuture;
+
+
 /**
  * A callback interface that users of PubSubProducerAdapter should implement if they want
  * to execute some code once PubSubProducerAdapter#sendMessage request is completed.
  */
-public interface PubSubProducerCallback {
+public abstract class PubSubProducerCallback extends CompletableFuture<PubSubProduceResult> {
```

Comment:
If I understand correctly how this works, it is the responsibility of the subclass to make sure in its implementation of the `onCompletion` function, it also calls either `CompletableFuture::complete` or `completeExceptionally`? For example, `ApacheKafkaProducerCallback` has this wiring.

Isn't this a little bit brittle though? I think this wiring should be part of the abstract class. The subclass could implement another function which also gets called by the abstract wiring. LMK what you think.

Reply:
> it is the responsibility of the subclass to make sure in its implementation of the onCompletion function, it also calls either CompletableFuture::complete or completeExceptionally

This is not the case. `ApacheKafkaProducerCallback` does not extend `PubSubProducerCallback` as it is part of the PubSub adapter. 

Marking future complete is the responsibility of the adapter.  The Kafka adapter delegates it to its callback.  

Here is the comment from `PubSubProducerAdapter::sendMessage`
```
   * @param pubSubProducerCallback   A CompletableFuture-come-callback which will be "invoked and completed" when
   *                                message is sent successfully.
   * All implementations of this interface MUST guarantee that "pubSubProducerCallback" will be
   *   1) invoked (call PubSubProducerCallback::onCompletion) and
   *   2) marked completed (PubSubProducerCallback::completeExceptionally or PubSubProducerCallback::complete).
   */
  void sendMessage(
  ```

Reply:
> Isn't this a little bit brittle though? I think this wiring should be part of the abstract class. The subclass could implement another function which also gets called by the abstract wiring. LMK what you thin

We can move it to the abstract class to avoid issues when the adapter implementation provider forgets to mark the future complete.